Initial commit

ThomasHakkers opened the pull request:
Essentially Torch ripped out of the original Zosma

ThomasHakkers committed:
Here be dragons

ThomasHakkers committed:
:sparkles: Added CircleCI badge

ThomasHakkers committed:
:sweat_smile: Add gradle wrapper properties

thumbnail commented on src/main/java/io/leonis/torch/Gradient.java:
```diff
@@ -0,0 +1,42 @@
+ * @author Thomas Hakkers
+ */
+public final class Gradient implements Function<Double, TextColor.RGB> {
+  private final List<Color> colors;
```
This implementation is broken and should be revisited. since it only takes the colors from the farthest colors apart; whilst it should support a linear gradient. for 2 colors a crossover at 50%, faded over the whole length. for 3 colors at 33% etc.

thumbnail commented on src/main/java/io/leonis/torch/MoveWindowBehaviour.java:
```diff
@@ -0,0 +1,34 @@
+import com.googlecode.lanterna.input.*;
+
+/**ss
+ * @author jeroen.dejong.
```
description

thumbnail commented on src/main/java/io/leonis/torch/MoveWindowBehaviour.java:
```diff
@@ -0,0 +1,34 @@
+      switch(keyStroke.getKeyType()) {
+        case ArrowDown:
+          target.setPosition(position.withRelative(new TerminalPosition(0, 1)));
+          return true;
```
we might not want to `return true;` because this prevents other listeners to be triggered.

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,41 @@
+  /**
+   * Constructs a new window manager.
+   */
+  public Torch(final Function<WindowBasedTextGUI, WindowBasedTextGUI> GUI) {
```
GUI shouldn't be all caps.

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,41 @@
+    try {
+      final Terminal terminal = new DefaultTerminalFactory(System.out, System.in, Charset.forName("UTF8")).createTerminal();
+      Screen screen = new VirtualScreen(new TerminalScreen(terminal));
+      screen.startScreen();
```
@ThomasHakkers You researched click behavior in the terminal. what should be done to add it ?

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,41 @@
+      textGUI.addListener(new CycleWindowBehaviour());
+      ((AsynchronousTextGUIThread)textGUI.getGUIThread()).start();
+    } catch (final Exception exception) {
+      throw new RuntimeException("Houston....", exception);
```
We could just make Torch raise the exception if we're just rethrowing it anyways.

thumbnail commented on src/main/java/io/leonis/torch/component/ZosmaBackground.java:
```diff
@@ -0,0 +1,51 @@
+          text.setForegroundColor(new TextColor.RGB(brighterColor.getRed(), brighterColor.getGreen(), brighterColor.getBlue()));
+        }
+        graphics.fill(' ');
+        // TODO this should be centered
```
This todo can be resolved in this PR.

thumbnail replied:
this is still applicable.

thumbnail commented on src/main/java/io/leonis/torch/component/ZosmaBackground.java:
```diff
@@ -0,0 +1,51 @@
+ *
+ * @author jeroen.dejong.
+ */
+@Value
```
@Value adds nothing here besides making the field `text` final. please remove `@Value` in favor of an explicit `final`

thumbnail commented on src/main/java/io/leonis/torch/component/ZosmaBackground.java:
```diff
@@ -0,0 +1,51 @@
+
+  public ZosmaBackground(TextColor color, String text) {
+    super(color);
+    this.text = new Label(text);
```
~If you create the Label in the renderer this constructor can be more lazy.~

thumbnail commented on src/main/java/io/leonis/torch/component/form/Form.java:
```diff
@@ -0,0 +1,85 @@
+  public synchronized List<FormElement<?>> getFormData() {
+    return this.elements.getChildren().stream()
+        .filter(FormElement.class::isInstance)
+        .map(child -> (FormElement<?>) child)
```
This can be changed into `map(FormElement<?>.class::cast)`

thumbnail commented on src/main/java/io/leonis/torch/component/form/FormBuilder.java:
```diff
@@ -0,0 +1,121 @@
+ *
+ * @author Jeroen de Jong
+ */
+public interface FormBuilder {
```
This might be refactored to be based on decorators and use overloading to prevent the types from being lost (and reflections being used). I'll do this separately.

thumbnail commented on src/main/java/io/leonis/torch/component/form/Form.java:
```diff
@@ -0,0 +1,85 @@
+   * @param buttons  the buttons
+   * @return the form
+   */
+  public static Form of(
```
This can be moved to a simple constructor.

thumbnail commented on src/main/java/io/leonis/torch/component/graph/LineGraph.java:
```diff
@@ -0,0 +1,55 @@
+ */
+@Value
+@AllArgsConstructor
+public class LineGraph extends AbstractComponent<LineGraph> {
```
This code needs verification.

thumbnail commented on src/main/java/io/leonis/torch/window/RamWindow.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * @author jeroen.dejong.
+ */
+public class RamWindow extends BasicWindow {
```
This class can be removed.

thumbnail commented on src/main/java/io/leonis/torch/window/Modal.java:
```diff
@@ -0,0 +1,34 @@
+   */
+  public Modal(final String title, final Component component) {
+    super(title);
+    this.setHints(Arrays.asList(Hint.MODAL, Hint.CENTERED));
```
This might be implemented using overloading; e.g. make `getHints` return the `super` with `Hint.MODAL` and `Hint.CENTERED` added by default.

thumbnail commented on src/main/java/io/leonis/torch/window/RxPanel.java:
```diff
@@ -0,0 +1,37 @@
+  public RxPanel addComponent(final Publisher<? extends Component> publisher) {
+    Flux.from(publisher)
+        .subscribeOn(Schedulers.single())
+        .transform(componentFlux -> {
```
maybe just solve this in two seperate fluxes; of which one flux will simply subscribes for the first; and the second has the buffer(2,1).

thumbnail reviewed (approved):
Not too much needs changing. I'll change a lot of these issues; we'll discuss how we can split these issues :)

thumbnail reviewed (changes requested):
I failed to check `request changes` last night. :')

thumbnail committed:
Add gitignore

thumbnail committed:
Remove field builder

thumbnail committed:
Remove unused dependencies

thumbnail committed:
Prevent swallowing IOExceptions in Torch

thumbnail committed:
Update gitignore

thumbnail committed:
Remove netty and ipc dependencies

thumbnail committed:
Remove RamWindow

thumbnail committed:
Rename behaviour to handler, moved to handlers package

thumbnail committed:
Created a GradientLine; refactored Gradient to work for just two colors

thumbnail committed:
Made LineType implement Function

thumbnail committed:
Cleaned ZosmaBackground up; made it more generic as well

thumbnail committed:
Make Modal a decorator

thumbnail committed:
Make LineGraph expect Functions instead of LineType's

thumbnail committed:
Make GradientLine depend on Function instead of custom interfaces

thumbnail committed:
Update GraphImage to work with new LineType and Gradient

thumbnail committed:
Call super in Equals, resloving remaining warning

ThomasHakkers committed:
Fix package name. Remove unused imports

ThomasHakkers committed:
:ambulance: Fix LineGraph

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,63 @@
+import com.googlecode.lanterna.terminal.Terminal;
+import io.leonis.torch.component.TextBackground;
+import io.leonis.torch.component.graph.Gradient;
+import io.leonis.torch.component.graph.LineGraph;
```
Remind me to provide you with the correct settings for intellij in slack. We decided to use a wildcard import if more than 1 dependency from the same package. (e.g. all `io.leonis.torch.component.graph.<specific>` imports should be changed to `io.leonis.torch.component.graph.*`). Luckily this is a setting and we can fix this during the final formatting.

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,63 @@
+
+/**
+ * The Class Torch.
+ *
```
docs

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author Jeroen de Jong
+ */
+public final class Torch {
```
Torch should implement Runnable, to make rendering on a seperate thread more simple.

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,63 @@
+    ((AsynchronousTextGUIThread)textGUI.getGUIThread()).start();
+  }
+
+  // TODO Remove. Currently here for debugging LineGraph
```
We might not want `SeparateTextGUIThread` because torch might run in a separate thread anyway. Not using `SeparateTextGUIThread` will allow the impl to wait for windows etc. :)

thumbnail commented on src/main/java/io/leonis/torch/Torch.java:
```diff
@@ -0,0 +1,63 @@
+  /**
+   * Constructs a new window manager.
+   */
+  public Torch(final Function<WindowBasedTextGUI, WindowBasedTextGUI> GUI) throws IOException {
```
This function might become a Consumer or smth. That way the side effecting isn't masked.

thumbnail replied:
The code in the constructor should be moved to a `#run` method as well though.

thumbnail commented on src/main/java/io/leonis/torch/component/graph/LineGraph.java:
```diff
@@ -0,0 +1,66 @@
+@Value
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper=true)
+public class LineGraph extends AbstractComponent<LineGraph> {
```
Needs to be merged with `GraphImage`

thumbnail committed:
Make Torch implement Runnable. Don't run lanterna on a seperate thread.

Torch can be ran in a seperate thread if desired

thumbnail committed:
Optimize imports

thumbnail committed:
Reformat code

thumbnail committed:
Combine GraphImage into LineGraph

thumbnail committed:
Add mouse capture mode.

If torch is run from the terminal it won't spawn a window; and clicking will work

thumbnail committed:
Fixed NPE

thumbnail committed:
Remove imports and remove null check. let it burn

thumbnail committed:
Add a mouseclickhandler

thumbnail committed:
Add ActiveWindowSelection. Add javadoc, Add DragWindowHandler

thumbnail committed:
Add explicit Listeners for mouse events and multiwindow guis

thumbnail committed:
Optimize CycleWindowHandler and MoveWindowHandler

thumbnail committed:
Move clicking on a window to a generic handler

thumbnail committed:
Split window dragging to specific classes

thumbnail committed:
Remove failed abstraction

thumbnail committed:
Move start of dragging to a class

thumbnail committed:
Rm -rf newline

thumbnail committed:
Only start dragging on CLICK_DOWN

thumbnail committed:
Remove extra newline

thumbnail commented:
Quick note to myself to remove the `static main` from `Torch` before merging.

thumbnail committed:
Finals and random code optimalisations

thumbnail committed:
Add javadoc for StartDragHandler

thumbnail committed:
Selfdestruct StopDragHandler once triggered

thumbnail committed:
Change author tags from jeroen.dejong. to Jeroen de Jong

thumbnail committed:
Add javadoc for all handlers

thumbnail committed:
Add javadoc for the windows

thumbnail committed:
Add final modifier to all classes

thumbnail committed:
Improve javadoc for TextBackground

thumbnail committed:
Update graph javadoc

thumbnail committed:
Reformat code

romni commented on build.gradle:
```diff
@@ -0,0 +1,30 @@
+
+dependencies {
+    // rx
+    compile 'io.projectreactor:reactor-core:3.0.5.RELEASE'
```
`reactor-core` is at `3.1.1.RELEASE`. could you bump the version? haven't checked the other dependencies but those might be updated as well

romni commented on src/main/java/io/leonis/torch/component/TextBackground.java:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  @Override
+  protected ComponentRenderer<EmptySpace> createDefaultRenderer() {
```
how often is this method called? it might be better to provide the below anonymous class as a static inner class in order to prevent dynamically extending ComponentRenderer with a new anonymous type every time the function is called

thumbnail replied:
Once per render call; but it's all cached internally afaik. It's usually done this way, [see lanterna](https://github.com/mabe02/lanterna/blob/master/src/main/java/com/googlecode/lanterna/gui2/EmptySpace.java#L87)

romni commented on src/main/java/io/leonis/torch/component/graph/LineGraph.java:
```diff
@@ -0,0 +1,98 @@
+    return data.stream().mapToDouble(Number::doubleValue).max().orElse(1);
+  }
+
+  private double getMin() {
```
some of these methods are missing javadoc

romni commented on src/main/java/io/leonis/torch/component/graph/LineGraph.java:
```diff
@@ -0,0 +1,98 @@
+
+  @Override
+  protected ComponentRenderer<LineGraph> createDefaultRenderer() {
+    return new ComponentRenderer<LineGraph>() {
```
same comment about anonymous classes applies here

thumbnail replied:
Same response about anonymous classes applies here.

romni commented on src/main/java/io/leonis/torch/handlers/CycleWindowHandler.java:
```diff
@@ -0,0 +1,24 @@
+
+  @Override
+  public boolean onUnhandledKeyStroke(final WindowBasedTextGUI textGUI, final KeyStroke keyStroke) {
+    if (keyStroke.getKeyType().equals(KeyType.Character) && "[]"
```
that `"[]"` should stick with the `.contains` below

thumbnail replied:
lel autoformatting; will fix.

romni commented on src/main/java/io/leonis/torch/handlers/MultiWindowListener.java:
```diff
@@ -0,0 +1,31 @@
+
+  @Override
+  default boolean onUnhandledKeyStroke(final TextGUI textGUI, final KeyStroke keyStroke) {
+    return textGUI instanceof WindowBasedTextGUI && onUnhandledKeyStroke(
```
i would break right before `&&` and keep  the call to `onUnhandledKeyStroke` with its arguments

thumbnail replied:
lel autoformatting; will fix.

thumbnail commented on src/main/java/io/leonis/torch/component/graph/LineGraph.java:
```diff
@@ -0,0 +1,98 @@
+      }
+
+      @Override
+      public void drawComponent(final TextGUIGraphics graphics, final LineGraph component) {
```
@romnous Any suggestions on improving this?

thumbnail committed:
Implement newlines

thumbnail committed:
Update lanterna, lombok and reactor to latest versions

romni commented on src/main/java/io/leonis/torch/component/graph/LineGraph.java:
```diff
@@ -0,0 +1,98 @@
+              if (number > 0d) {
+                IntStream.rangeClosed(xAxisRow, (int) Math.ceil(rowValue))
+                    .forEach(row -> {
+                      final double ratio = ((double) getRows() - row) / ((double) getRows() - 0);
```
`- 0`? :confused:

romni commented on src/main/java/io/leonis/torch/component/graph/LineGraph.java:
```diff
@@ -0,0 +1,98 @@
+              } else {
+                IntStream.rangeClosed((int) Math.floor(rowValue), xAxisRow)
+                    .forEach(row -> {
+                      final double ratio = ((double) getRows() - row) / ((double) getRows() - 0);
```
`- 0`??

thumbnail committed:
Add this everywhere

sneaked a fix for flux deprecation in as well

ThomasHakkers committed:
Remove unnecessary subtraction of 0 and casting to double

thumbnail committed:
Remove empty constructor

thumbnail committed:
Remove TextBackground

thumbnail committed:
Remove TODO

thumbnail committed:
Remove example code
